src/dmd/backend/dwarfdbginf.d

```diff
@@ -1353,50 +1435,135 @@ static if (1)
+                    }
+
+                    if (last_filename == filename)
+                    {
+                        ld.filenumber = last_filenumber;
+                    }
                     else
-                        filename = .filename;
+                    {
+                        ld.filenumber = dwarf_line_addfile(filename);
+                        dwarf_line_add_directory(filename);
+
+                        last_filenumber = ld.filenumber;
+                        last_filename = filename;
+                    }
                 }
+            }
 
-                if (last_filename == filename)
+            if (config.dwarf >= 5)
+            {
+                debug_line.buf.writeuLEB128(lineDirectories.length());   // directories_count
+            }
+            // include_directories
+            auto dirkeys = lineDirectories.aa.keys();
+            if (dirkeys)
+            {
+                foreach (id; 1 .. lineDirectories.length() + 1)
                 {
-                    ld.filenumber = last_filenumber;
+                    foreach (const(char)[] dir; dirkeys)
+                    {
+                        // Directories must be written in the correct order, to match file_name indexes
+                        if (dwarf_line_add_directory(dir) == id)
+                        {
+                            //printf("%d: %s\n", dwarf_line_add_directory(dir), dir);
+                            debug_line.buf.writeString(dir);
+                            break;    
+                        }
+                    }
                 }
-                else
+                free(cast(void*) dirkeys);
+                dirkeys = null;
+            }
+
+            if (config.dwarf < 5)
+            {
+                debug_line.buf.writeByte(0);   // end of include_directories
+            }
+            else
+            {
+                struct FileNameEntryFormat
                 {
-                    ld.filenumber = dwarf_line_addfile(filename);
+                    ubyte count = format.sizeof / 2;
+                    ubyte[4] format =
+                    [
+                        DW_LNCT_path,               DW_FORM_string,
+                        DW_LNCT_directory_index,    DW_FORM_data1,
+                    ];
+                }
+                auto file_name_entry_format = FileNameEntryFormat.init;
+                debug_line.buf.write(&file_name_entry_format, file_name_entry_format.sizeof);
 
-                    last_filenumber = ld.filenumber;
-                    last_filename = filename;
+                debug_line.buf.writeuLEB128(infoFileName_table.length);  // file_names_count
+            }
+                
+
+            // file_names
+            auto filekeys = infoFileName_table.aa.keys();
+            if (filekeys)
+            {
+                foreach (id; 1 .. infoFileName_table.length() + 1)
+                {
+                    foreach (const(char)[] filename; filekeys)
```

Comment:
If someone has a way to get every key in the correct order (order in which they were added), I'm taking.